addons/web/static/tests/legacy/fields/basic_fields_tests.js

```diff
@@ -1318,6 +1318,23 @@ QUnit.module('basic_fields', {
 
         var charField = _.find(form.renderer.allFieldWidgets)[0];
         assert.strictEqual(charField.isValid(), true);
+
+        this.data.partner.fields.foo.pattern = ['Only Belgian keyboard', 'azerty'];
+        var form = await createView({
+            View: FormView,
+            model: 'partner',
+            data: this.data,
+            arch:'<form string="Partners">' +
+                        '<field name="foo"/>' +
+                '</form>',
+            res_id: 1,
+        });
+        assert.strictEqual(_.find(form.renderer.allFieldWidgets)[0].isValid(), false);
+
+        await testUtils.form.clickEdit(form);
+        await testUtils.fields.editInput(form.$('input[type="text"].o_field_widget'), 'azerty');
+        await testUtils.form.clickSave(form);
+        assert.strictEqual(_.find(form.renderer.allFieldWidgets)[0].isValid(), true);
```

Comment:
you should check the DOM instead: the field should be marked as invalid (classname o_invalid or something) + no rpc should have been done
We try to avoid whitebox assertions as much as possible, and rather assert the "observable state"